---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# BACDAC

<!-- badges: start -->
<!-- badges: end -->

BACDAC is an R package for "Ploidy Analysis using the Binomial Allelic Content with 
Discretization Algorithm and Constellation plot"

Reports tumor ploidy and purity from whole-genome sequencing data including low-pass low-tumor NGS.

Inputs include read-depth and segmentation data, and ref/alt counts for common single nucleutide 
polymorphisms (SNPs). Calculates the "Heterozygosity Score" and produces the Constellation plot 
to visualize allele-specific copy-number as shown in publication xyz.


## Installation

install BACDAC using R: `devtools::install_github('vasmatzis/BACDAC')`


## Example

This is a quick example to confirm you can load example data from the package and use one of the 
simple functions. A successful attempt will generate a linear genome plot similar to figure 3 of 
the publication:

```{r example}
## try to run this example:
  library(BACDAC)
  library(logging)

  basicConfig("DEBUG")
  sampleId='TCGA-14-1402-02A_ds'; alternateId=66301

  # inputDir: path to the package example data
  inputDir <- system.file('extdata', package = "BACDAC")
  segmentationFile <- file.path(inputDir, paste0(sampleId, '_segmentation.csv'))
  segmentation= read.csv(segmentationFile, comment.char = '#', header = TRUE) 
  # check for columns: chr, start, end, rd per segment
  segmentation=checkSegmentation(segmentation)
  thirtyKbFile=file.path(inputDir, paste0(sampleId,'_','readDepthPer30kbBin.Rds'))
  readDepthPer30kbBin = readRDS(file=thirtyKbFile )
  readDepthBinSize=readDepthPer30kbBin$windowSize

  op <- par(mfrow=c(3,1),mai=c(.25,0.5, 0.3,0.25), mgp=c(2, .5, 0))
 # default cnv color coding and annotations
 linearGenomePlot(readDepthPer30kbBin=readDepthPer30kbBin,readDepthBinSize=readDepthBinSize, 
                  sampleId=sampleId,segmentation=segmentation)
```

There are four main functions to be operated in this order:

1) `calculateHetScore()`
2) `calculatePloidy()`
3) `loadStarsInTheClouds()`
4) `plotStarsInTheClouds()`

The function `runBACDAC()` will run all steps sequentially. See the example for help running this 
function. 

## Reference files
Two reference files are also required. These are available at https://zenodo.org/records/13619655  
1) hsNormMat 
2) testVals 
When running the examples (as packaged with their current directory structure, if the files do not
exist, they will be automatically downloaded and installed one level up from your current directory
into `../referencefiles`


## Sample input Files
There are four types of input files required for each sample you wish to test. The contents and any
required file naming scheme are described below. Example data is included with the package for
testing and illustration of the expected structure and format of the data. This data also allows for
full demonstration of the BACDAC features. The included data is from sample TCGA-14-1402-02A which
was down-sampled to a base coverage of 5x and then processed through our pipelines BIMA and
svaTools. We gave the sample the sampleId: `TCGA-14-1402-02A_ds`. All the input files are located in
a single directory, the `inputDir`. The required infput files are as follows:

1) refAltCount data:

23 files, one file for each chromosome 1-22 and X. The file is a matrix with four columns: chr pos
ref alt. The file must be in `inputDir` and the file name is of the form `<sampleId>_1.Rds`, saved
using `saveRDS()`

```{r refAltCount}
# example files in the package
inputDir <- system.file('extdata', package = "BACDAC")
sampleId='TCGA-14-1402-02A_ds';

dir(inputDir, pattern = 'refAltCount')
# load
refAltCountchr1=readRDS(file.path(inputDir, paste0(sampleId, '_refAltCount_1.Rds')))
# preview
head(refAltCountchr1)
str(segmentation)

```

2) segmentation data:  
One file with identified regions of the genome with constant read depth.   
A data frame with required columns: chr, start, end, rd; where `rd` is the read depth or read count
for that segment. 
optional column: cnvState (1=loss, 2=normal, 3=gain) for color coding the linear genome plot.  
The bin size used for creating the segmentation file will also be needed.

```{r segmentation}
# example files in the package
inputDir <- system.file('extdata', package = "BACDAC")
sampleId='TCGA-14-1402-02A_ds';
# file name
segmentationFile= file.path(inputDir, paste0(sampleId, '_segmentation.csv'))
# load
segmentation_org=read.csv(segmentationFile, comment.char = "#", header=TRUE)
# checks for and returns the required columns
segmentation=BACDAC::checkSegmentation(segmentation_org)
# preview
head(segmentation)
str(segmentation)
```


3)  readDepthPer30kbBin       
A list with two equal length arrays and one item:
  * `readDepthArray` read depth for 30 kb bins, normalized for GC content and other artifacts
  * `goodWindowArray` linear genome position of each window (bin), masked windows have been removed
  * `readDepthBinSize` size of the bins in the read depth array, expecting 30000 or 100000

4)  readDepthPer100kbBin       
list with two equal length arrays and one item:
  * `readDepthArray` read depth for 100 kb bins, normalized for GC content and other artifacts
  * `goodWindowArray` linear genome position of each window (bin) (masked windows have been removed)
  * `readDepthBinSize` size of the bins in the read depth array, expecting 30000 or 100000

```{r readDepthData}
inputDir <- system.file('extdata', package = "BACDAC")
sampleId='TCGA-14-1402-02A_ds';
dir(inputDir, pattern = 'readDepthPer')

thirtyKbFile=file.path(inputDir, paste0(sampleId,'_','readDepthPer30kbBin.Rds'))
#load
readDepthPer30kbBin = readRDS(file=thirtyKbFile )
# preview
names(readDepthPer30kbBin)
str(readDepthPer30kbBin)

hundredKbFile=file.path(inputDir, paste0(sampleId,'_','readDepthPer100kbBin.Rds'))
#load
readDepthPer100kbBin = readRDS(file=hundredKbFile )
# preview
names(readDepthPer100kbBin)
str(readDepthPer100kbBin)

```

## Linear Genome

Linear genome coordinates are used throughout for ease of indexing. Helper functions are provided 
to help transform from single bp scale to a desired bin size.

```{r}
coords=BACDAC::getLinearCoordinates()
# linear coordinate for the start of each chromosome where 23=X, 24=Y
coords@chromStart[1:24]

# linear coordinate for the start of each chromosome using 30 kb bins
BACDAC::binnedPosStart(coords@chromStart[1:24], binSize=30000)

```



